File: android/src/main/kotlin/com/bruno/native_admob_flutter/banner/Controller.kt
package com.bruno.native_admob_flutter.banner

import android.content.Context
import com.google.android.gms.ads.AdView
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

class BannerAdController(
        val id: String,
        val channel: MethodChannel,
        val context: Context
) : MethodChannel.MethodCallHandler {

    var adView: AdView? = null

    /// New native ad when loaded
    var loadRequested: ((MethodChannel.Result) -> Unit)? = null

    init {
        channel.setMethodCallHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "loadAd" -> {
                channel.invokeMethod("loading", null)
                loadRequested?.let { it(result) }
                // Actually, no ads will be loaded here, so return false immediately
                result.success(false)
            }
            else -> result.notImplemented()
        }
    }

}

object BannerAdControllerManager {
    private val controllers: ArrayList<BannerAdController> = arrayListOf()

    fun createController(id: String, binaryMessenger: BinaryMessenger, context: Context) {
        if (getController(id) == null) {
            val methodChannel = MethodChannel(binaryMessenger, id)
            val controller = BannerAdController(id, methodChannel, context)
            controllers.add(controller)
        }
    }

    fun getController(id: String): BannerAdController? {
        return controllers.firstOrNull { it.id == id }
    }

    fun removeController(id: String) {
        val index = controllers.indexOfFirst { it.id == id }
        if (index >= 0) controllers.removeAt(index)
    }
}

File: ios/Classes/Banner/Controller.swift
import Flutter
import GoogleMobileAds

class BannerAdController: NSObject {

    var bannerView: GADBannerView!

    var loadRequested: ((MethodChannel.Result) -> Unit)? = null

    let id: String
    let channel: FlutterMethodChannel

    init(id: String, channel: FlutterMethodChannel) {
        self.id = id
        self.channel = channel
        super.init()
        
        channel.setMethodCallHandler(handle)
    }

    private func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let params = call.arguments as? [String: Any]
        
        switch call.method {
            case "loadAd":
                channel.invokeMethod("loading", arguments: nil)
            default:
                result(FlutterMethodNotImplemented)
        }
    }

}

class BannerAdControllerManager {
    
    static let shared = BannerAdControllerManager()
    
    private var controllers: [BannerAdController] = []
    
    private init() {}
    
    func createController(forID id: String, binaryMessenger: FlutterBinaryMessenger) {
        if getController(forID: id) == nil {
            let methodChannel = FlutterMethodChannel(name: id, binaryMessenger: binaryMessenger)
            let controller = BannerAdController(id: id, channel: methodChannel)
            controllers.append(controller)
        }
    }
    
    func getController(forID id: String) -> BannerAdController? {
        return controllers.first(where: { $0.id == id })
    }
    
    func removeController(forID id: String) {
        if let index = controllers.firstIndex(where: { $0.id == id }) {
            controllers.remove(at: index)
        }
    }
}